### PR TITLE
Feature/android pip

### PR DESCRIPTION
**Pull request by matejpolak**
## Provide an example of how to test the change
In order to test this change, you should follow all new steps listed in readme under picureInPicture props:

* Add following to the AndroidManifest.xml -> MainActivity:
```
  android:configChanges="keyboard|keyboardHidden|orientation|screenSize|screenLayout"
  android:resizeableActivity="true"
  android:supportsPictureInPicture="true"
```
* [Enable PIP feature for you app](https://support.google.com/youtube/answer/7552722?co=GENIE.Platform%3DAndroid&hl=en) in your device settings
* Add following methods to MainActivity:
```
    // To let JS knows about PIP mode change. 
    override fun onPictureInPictureModeChanged(isInPictureInPictureMode: Boolean, newConfig: Configuration?) {
       super.onPictureInPictureModeChanged(isInPictureInPictureMode, newConfig)
       val intent = Intent("onPictureInPictureModeChanged")
       intent.putExtra("isInPictureInPictureMode", isInPictureInPictureMode)
       this.sendBroadcast(intent)
    }
    
    // To trigger PIP mode by pressing `Home` or `Recent` buttons
    override fun onUserLeaveHint() {
       val intent = Intent("onUserLeaveHint")
       this.sendBroadcast(intent)
       super.onUserLeaveHint()
    }
        
    // To close and kill the app when closing PIP window.
    override fun onStop() {
       super.onStop()
       finishAndRemoveTask()
    }
``` 

### Describe the changes
So far PIP mode has been supported only on iPad. From now on, it should be working also on all Android Phones and Tablets running Oreo at least. This implementation includes:

- Ability to toggle PIP mode by setting `pictureInPicture` prop.
- `onPictureInPictureStatusChanged` prop callback returning PIP state same as iOS does.
- Ability to turn on PIP mode by pressing Home/Recent HW button.
- Previous / Pause/Play / Next PIP window buttons.
- Implementation of closing and killing the entire app when closing PIP window.

android-PIP

### CODE REVIEW

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 1]
```diff
@@ -179,6 +189,30 @@ public ReactExoplayerView(ThemedReactContext context, ReactExoplayerConfig confi
...
+            public void onReceive(Context context, Intent intent) {
+                self.setPictureInPicture(true);
+
+                // Manually call onHostResume to keep video playing.
+                self.onHostResume();
+            }
+        };
+
+        Activity activity = themedReactContext.getCurrentActivity();
+        activity.registerReceiver(pipReceiver, new IntentFilter("onPictureInPictureModeChanged"));
```
Comment (benoitdion) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:212]: should we unregister the receivers?

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 2]
```diff
@@ -179,6 +189,30 @@ public ReactExoplayerView(ThemedReactContext context, ReactExoplayerConfig confi
         themedReactContext.addLifecycleEventListener(this);
         audioBecomingNoisyReceiver = new AudioBecomingNoisyReceiver(themedReactContext);
 
+        ReactExoplayerView self = this;
+
+        BroadcastReceiver pipReceiver = new BroadcastReceiver() {
```
Comment (benoitdion) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: can this be an optional behavior? maybe props based?

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 3]
```diff
@@ -398,6 +432,12 @@ public void run() {
                 initializePlayerControl();
                 setControls(controls);
                 applyModifiers();
+
+                //Use Media Session Connector from the ExoPlayer library to enable MediaSession Controls in PIP.
+                MediaSessionCompat mediaSession = new MediaSessionCompat(getContext(), "tag");
```
Comment (benoitdion) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: maybe media session management should be its own feature, independent of PIP support?

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 4]
```diff
@@ -398,6 +432,12 @@ public void run() {
...
+                MediaSessionConnector mediaSessionConnector = new MediaSessionConnector(mediaSession);
+                mediaSessionConnector.setPlayer(player, null);
+                mediaSession.setActive(true);
```
Comment (benoitdion) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: we may need to clean up the session later in the lifecycle.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 2 above)
Comment (matejpolak) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Of course, good idea!

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 3 above)
Comment (matejpolak) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Yeah, could be of course. Should I make another PR for the standalone media session management?

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 2 above)
Comment (matejpolak) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Well, actually it's an essential part of the implementation which is required to make it work. In this case, pipReceiver is responsible for PIP mode changes, so if user returns to the regular app (press maximise button), this receiver notifies videoEventEmitter which afterwards trigger onPictureInPictureStatusChanged callback to notify JS about such a change. Without this, client would never be able to detect whether the app is running in PIP mode or not.

But I can of course create the other receiver prop driven. That one is responsible for triggering PIP mode when user press Home or Recent button, which should be optional indeed.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 3 above)
Comment (matejpolak) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: https://github.com/react-native-community/react-native-video/pull/1780

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 4 above)
Comment (matejpolak) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: https://github.com/react-native-community/react-native-video/pull/1780

**Comment by dylancom** [comment]
Why isn't this merged yet?

**Comment by Spoutnik97** [comment]
Hi !
Is there an update of the pull request ?

**Comment by matejpolak** [comment]
Hello everyone, let me take a look at this PR again, solve the conflicts, and then we are ready for reviews and eventually for merging. Anyways I would need someone to give a green light.

I think I should be able to get back to this somewhere this week.

**Comment by mbrimmer83** [comment]
Hi,
Any updates on this pull request?
Thanks

**Comment by michael-kiley-verified** [comment]
Hi all,
This looks like good work and is a feature I would really love to get access to. Is this still active?
Thanks!

**Comment by KingAmo** [comment]
@nickfujita what do you think about this pr

**Comment by victoromarques** [comment]
How do I apply this change to my project? how do i install this update?

**Review by uranashel44 — APPROVED** [review]

**Review by uranashel44 — APPROVED** [review]

**Comment by TomMichalkevic** [comment]
Any updates on this PR?

**Comment by jrhager84** [comment]
How do we implement this in our react-native-video projects? It seems to be great. Why hasn't this been merged?

**Comment by hueniverse** [comment]
I automatically closed PRs and issues that were stale. We can reopen if there is interest and resources to review and test. If you are interested in helping getting this PR through, I'm happy to reopen. Merging this would require 3 people to review, test, and approve it.

**Comment by jrhager84** [comment]
> I automatically closed PRs and issues that were stale. We can reopen if there is interest and resources to review and test. If you are interested in helping getting this PR through, I'm happy to reopen. Merging this would require 3 people to review, test, and approve it.

I would, actually. I'll see if I can carve some time. 😎👍

**Comment by freeboub** [comment]
Code looks good, but I am afraid this change is not really expected in react-native-video...
I agree this is a must have feature ! :) 
But pip is not be linked to react-native-video (google maps also supports pip)
I have just publish a fork of rn-android-pip here: https://github.com/iFeelSmart/rn-android-pip/
With this project I am able to use pip in any app (including react-native-video).

For exemple, if tomorrow you want to do pip with react-native-youtube, (https://www.npmjs.com/package/react-native-youtube) you can have issues to make multiple pip handling bindings ...

**Comment by jrhager84** [comment]
> Code looks good, but I am afraid this change is not really expected in react-native-video... I agree this is a must have feature ! :) But pip is not be linked to react-native-video (google maps also supports pip) I have just publish a fork of rn-android-pip here: https://github.com/iFeelSmart/rn-android-pip/ With this project I am able to use pip in any app (including react-native-video).
> 
> For exemple, if tomorrow you want to do pip with react-native-youtube, (https://www.npmjs.com/package/react-native-youtube) you can have issues to make multiple pip handling bindings ...

What about pip for ios? Wouldn't react native youtube support that as well? In my mind it would make sense to handle pip in rn video because it also supports ios.

**Comment by freeboub** [comment]
I agree that ios pip is already handle by react-native-video so it make sens to support it on android.
For react-native-youtube, i don.t think they support...
I just want to highlight it can cause conflicts with other implementations...

**Comment by freeboub** [comment]
In other words, it is very useful for quick integration, but it doesn't fullfil all expected requirements of an advanced android application :)

**Comment by jrhager84** [comment]
> In other words, it is very useful for quick integration, but it doesn't fullfil all expected requirements of an advanced android application :)

What advanced requirements are you referring to? Is it something that other package can do? Maybe I just don't know enough about android pip to weigh in.

**Comment by TomMichalkevic** [comment]
I'll make some time to test this tomorrow ;) Let's get this reviewed and merged.

**Comment by TomMichalkevic** [comment]
So I've tested it a bit and I'm not sure it works the way it should:
image
The app crashes now each time there is an attempt to play any video with the added pictureInPicture prop.
If I remove the prop, it works fine. As in, no PiP, but playback works again.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 5]
```diff
@@ -1212,4 +1254,49 @@ public void setControls(boolean controls) {
...
+    /**
+     * PIP handled, for N devices that support it, not "officially".
+     */
+    public void enterPictureInPictureMode() {
+        PackageManager packageManager = themedReactContext.getPackageManager();
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N
+                && packageManager
+                .hasSystemFeature(
+                        PackageManager.FEATURE_PICTURE_IN_PICTURE)) {
+            long videoPosition = player.getCurrentPosition();
```
Comment (TomMichalkevic) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:1288]: This line should be removed. The variable is not used anywhere and causes crashes.

**Comment by TomMichalkevic** [comment]
With the line removed above it continues on working, but the PiP floating window disappears.
Here's what I get from logcat:
`MainActivity has leaked window DecorView@ca3e234[] that was originally added here at android.view.ViewRootImpl.<init>(ViewRootImpl.java:733)
        at android.view.ViewRootImpl.<init>(ViewRootImpl.java:717)
        at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:399)
        at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:109)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4535)
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)`

**Comment by hueniverse** [comment]
@TomMichalkevic Thanks for the review.

@matejpolak Are you still interested in getting this meged?

**Comment by paulohc** [comment]
Hello @hueniverse!
Is there still any chance to reopen and merge this PR?

**Comment by hueniverse** [comment]
@PauloHC there is no point to reopen it because it needs to be rebased. If you want to do the work and open a new PR with these changes against master, it will get reviewed. Someone needs to do the work.

**Comment by watadarkstar** [comment]
Here is the Java code for anyone who needs it as the instructions were written in Kotlin for the MainActivity:

```
import android.content.Intent;
import android.content.res.Configuration;

...
    
@Override
protected void onStop() {
    super.onStop();
    finishAndRemoveTask();
}

@Override
protected void onUserLeaveHint() {
    super.onUserLeaveHint();
    Intent intent = new Intent("onUserLeaveHint");
    this.sendBroadcast(intent);
}

@Override
public void onPictureInPictureModeChanged(boolean isInPictureInPictureMode, Configuration newConfig) {
    super.onPictureInPictureModeChanged(isInPictureInPictureMode, newConfig);
    Intent intent = new Intent("onPictureInPictureModeChanged");
    intent.putExtra("isInPictureInPictureMode", isInPictureInPictureMode);
    this.sendBroadcast(intent);
}
```

**Pull request closed without merging**